Firestore/core/src/firebase/firestore/local/memory_lru_reference_delegate.cc

```diff
@@ -0,0 +1,190 @@
+  QueryData updated =
+      queryData.Copy(queryData.snapshot_version(), queryData.resume_token(),
+                     current_sequence_number_);
+  persistence_->query_cache()->UpdateTarget(updated);
+}
+
+void MemoryLruReferenceDelegate::UpdateLimboDocument(
+    const model::DocumentKey& key) {
+  sequence_numbers_[key] = current_sequence_number_;
+}
+
+void MemoryLruReferenceDelegate::OnTransactionStarted(absl::string_view label) {
+  current_sequence_number_ = listen_sequence_->Next();
+}
+
+void MemoryLruReferenceDelegate::OnTransactionCommitted() {
+  current_sequence_number_ = kListenSequenceNumberInvalid;
+}
+
+void MemoryLruReferenceDelegate::EnumerateTargets(
+    const TargetCallback& callback) {
+  return persistence_->query_cache()->EnumerateTargets(callback);
+}
+
+void MemoryLruReferenceDelegate::EnumerateOrphanedDocuments(
+    const OrphanedDocumentCallback& callback) {
+  for (const auto& entry : sequence_numbers_) {
+    ListenSequenceNumber sequenceNumber = entry.second;
+    const DocumentKey& key = entry.first;
+    // Pass in the exact sequence number as the upper bound so we know it won't
+    // be pinned by being too recent.
+    if (!IsPinnedAtSequenceNumber(sequenceNumber, key)) {
+      callback(key, sequenceNumber);
+    }
+  }
+}
+
+size_t MemoryLruReferenceDelegate::GetSequenceNumberCount() {
+  size_t totalCount = persistence_->query_cache()->size();
+  EnumerateOrphanedDocuments(
+      [&totalCount](const DocumentKey& key,
+                    ListenSequenceNumber sequenceNumber) { totalCount++; });
+  return totalCount;
+}
+
+int MemoryLruReferenceDelegate::RemoveTargets(
+    model::ListenSequenceNumber sequence_number,
+    const LiveQueryMap& live_queries) {
+  return persistence_->query_cache()->RemoveTargets(sequence_number,
+                                                    live_queries);
+}
+
+int MemoryLruReferenceDelegate::RemoveOrphanedDocuments(
+    model::ListenSequenceNumber upper_bound) {
+  std::vector<DocumentKey> removed =
+      persistence_->remote_document_cache()->RemoveOrphanedDocuments(
+          this, upper_bound);
+  for (const auto& key : removed) {
+    sequence_numbers_.erase(key);
+  }
+  return static_cast<int>(removed.size());
+}
+
+void MemoryLruReferenceDelegate::AddReference(const DocumentKey& key) {
+  sequence_numbers_[key] = current_sequence_number_;
+}
+
+void MemoryLruReferenceDelegate::RemoveReference(const DocumentKey& key) {
+  sequence_numbers_[key] = current_sequence_number_;
+}
+
+bool MemoryLruReferenceDelegate::MutationQueuesContainKey(
+    const DocumentKey& key) {
+  const auto& queues = persistence_->mutation_queues();
+  for (const auto& entry : queues) {
+    if (entry.second->ContainsKey(key)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void MemoryLruReferenceDelegate::RemoveMutationReference(
+    const DocumentKey& key) {
+  sequence_numbers_[key] = current_sequence_number_;
```

Comment:
Same question.

Reply:
Sorry, I meant to double-check that the implementation is correct (it duplicates other `Remove` functions). From your comment above, I believe it is.